# SMS Spam Collection Classifier


### Introduction
This project aims to build a classification model that is able to distinguish between spam and non-spam messages in SMS messages.

The dataset is obtained from Kaggle and created by UCI Machine Learning : [https://www.kaggle.com/datasets/uciml/sms-spam-collection-dataset](https://www.kaggle.com/datasets/uciml/sms-spam-collection-dataset)


First, I import the necessary Python libraries/packages. Additional imports (e.g. scikit-learn modules) will be added as needed throughout the project.

In [3]:
import pandas as pd
import numpy as np

### Goal 1: Load and Explore the Data
Load the CSV file into a pandas DataFrame and explore its structure such as columns, shape, data types, missing values, and class distribution.

In [4]:
# Load CSV file into pandas DataFrame
df = pd.read_csv('spam.csv', encoding='latin-1') # Using 'latin-1' encoding as the dataset contains characters incompatible with UTF-8

df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


Here, we can see a few columns, v1, v2, and some other unnamed columns. Now, I want to get rid of the unnamed columns as they provide no information, as well as rename columns v1 and v2 into something more meaningful.

v1 is renamed into "Classification" and v2 is renamed into "Content".

In [5]:
# Drop unnecessary columns
df = df.drop(columns=['Unnamed: 2','Unnamed: 3','Unnamed: 4'])
df.head()

,v1,v2
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [6]:
# Rename columns to be meaningful
df = df.rename(columns={'v1':'Classification', 'v2':'Content'})
df.head()

,Classification,Content
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


Next, I check the shape, data types, and missing values of the dataset.

In [7]:
print(df.shape)
print() 
print(df.dtypes)
print() 
print(df.isnull().sum())

(5572, 2)

Classification    object
Content           object
dtype: object

Classification    0
Content           0
dtype: int64


This dataset contains 5572 rows and 2 columns after removing irrelevant columns.

The datatypes for both columns are *object*, which is equivalent to string data.

Upon checking the count of missing values for each columns, the dataset appears to have all rows and columns filled, there are no empty values.

Next, I examine the distribution of spam vs ham to check for class imbalance.

In [8]:
df['Classification'].value_counts()

Classification
ham     4825
spam     747
Name: count, dtype: int64

This is an imbalanced dataset that is skewed towards a larger amount of non-spam data. Roughly 87% of these messages are intended for the recipient.

### Goal 2: Text Preprocessing, Feature Extraction, and Train/Test Split

Before building a model, the text data needs to be converted into numerical features. I will use scikit-learn's *TfidfVectorizer* to handle both basic text cleaning (lowercasing, stopword removal) and feature extraction (TF-IDF scores) in one step.

The data is split into training and test sets before fitting the TF-IDF vectorizer to avoid data leakage, ensuring the model is never exposed to patterns from the test set during training.

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer


In [ ]:
# Define X and y as features and labels respectively
X = df['Content']
y = df['Classification']

# Splits data into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Initialize vectorizer which normalizes words to lowercase and removes stopwords
vectorizer = TfidfVectorizer(lowercase=True, stop_words='english')

# Fit vectorizer on training data and transform into TF-IDF features
X_train_tfidf = vectorizer.fit_transform(X_train)

# Only transform test data using to avoid data leakage
X_test_tfidft = vectorizer.transform(X_test)
